### PR TITLE
discoverd2: Implement DNS server

### PR DESCRIPTION
**Pull request by titanous**
Incremental patch for #145.

This implements a DNS server backed by the discoverd database.

Queries that are not for the `discoverd.` domain will be proxied to a recursive nameservers if configured.

Each registered service exposes a number of domains:

#### Service

`SERVICE.discoverd.` returns randomly sorted A, AAAA, and SRV records for each instance.

#### RFC 2782

`_SERVICE._PROTO.discoverd.` returns randomly sorted A, AAAA, and SRV records for each instance, filtered by protocol.

#### Leader

`leader.SERVICE.discoverd.` returns single A or AAAA and SRV records pointing at the leader of the service.

#### Instance

`INSTANCE_ID.SERVICE._i.discoverd.` returns single A or AAAA and SRV records for a single instance. This domain pattern is typically only used as the target for SRV records (which include the port but not the IP address).

Lookups over UDP are limited to up to three instances and the response does not have the truncation bit set (which would trigger a subsequent lookup over TCP). TCP lookups return all instances. The weight and priority of SRV records is always 1. The response to SRV questions over TCP include extra A/AAAA records matching each target in the answer. The TTL is always 0.

### CODE REVIEW

File: discoverd2/server/dns.go [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+	}
+
+	api := dnsAPI{srv}
+	mux := dns.NewServeMux()
+	mux.HandleFunc(srv.Domain, api.ServiceLookup)
+	if len(srv.Recursors) > 0 {
+		mux.HandleFunc(".", api.Recurse)
+	}
+
+	errors := make(chan error, 4)
```
Comment (heavenlyhash) [inline discoverd2/server/dns.go]: 4?

Comment (titanous) [inline discoverd2/server/dns.go]: Two `nil`s from `done()` plus two errors from `ListenAndServe`, probably overkill but I didn't want to risk leaking goroutines.

Comment (heavenlyhash) [inline discoverd2/server/dns.go]: I guess this has something to do with 2 servers x 2 statuses, but I think this would make more sense as a channel per server, then selecting over them.

Comment (titanous) [inline discoverd2/server/dns.go]: A channel per server would require more code, as there would be a different `NotifyStartedFunc` for each one, and a for loop wrapping a select, is it really that hard to follow the current code?

Comment (heavenlyhash) [inline discoverd2/server/dns.go]: I didn't follow what sending the nils into the errors channel does?  It looks like the range just eats them, and isn't waiting on them to trigger anything?

Comment (titanous) [inline discoverd2/server/dns.go]: There are two places where a send can occur for each server:

- From the `NotifyStartedFunc` (`nil`) - this indicates that the server has started correctly
- From the goroutine running `server.ListenAndServe()` - this indicates that the server failed to start or is stopping

If the server fails to start, `NotifyStartedFunc` will not be called, and we'll get an error sent on `errors`. The range receives `len(srv.servers)` times from `errors` and returns an error if one is received.

File: discoverd2/server/dns_test.go [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+			domain: "leader.a.discoverd.",
+			data:   simpleData,
+			qs:     leaderQs,
+		},
+		{
+			name:   "v6 leader",
+			domain: "leader.a.discoverd.",
+			data:   v6v4Data[:1],
+			qs:     v6LeaderQs,
+		},
```
Comment (heavenlyhash) [inline discoverd2/server/dns_test.go]: Do any of these test datasets hit the recursors?

(I guess that's a hard area to reach... if we can't swallow the network dependency, maybe mocked recursors?)

Comment (titanous) [inline discoverd2/server/dns_test.go]: https://github.com/flynn/flynn/pull/746/files#diff-7b854941b569c25c17fd931d5e47c443R52

And they hit real servers as I wanted to be sure that it works.

**Comment by heavenlyhash** [comment]
LGTM
